local_planner/src/nodes/local_planner.cpp

```diff
@@ -279,13 +280,24 @@ void LocalPlanner::determineStrategy() {
           star_planner_->setReprojectedPoints(reprojected_points_,
                                               reprojected_points_age_);
           star_planner_->setCloud(final_cloud_);
+
+          // set last chosen direction for smoothing
+          PolarPoint last_wp_pol = cartesianToPolar(
```

Comment:
The direction vector from last_waypoint to current position should be:
PolarPoint last_wp_pol = cartesianToPolar( position,last_sent_waypoint);

Reply:
I think this line is correct as it is. last_wp_pol is not meant to be a direction vector from last_waypoint to current position. But a polar representation of the last wp in coordinate system with current pos as an origin

Reply:
"a polar representation of the last wp in coordinate system with current pos as an origin" that mean it's a direction vector from current pos ----> last wp. As i understand, if need smoothing the path, it should following last direction: that's last wp ---> current pos.

Reply:
@baumanta the origin is current pos, polar representation of the last wp, it's  the vector from current pos to last wp.

Reply:
Thanks for questioning the logic - that's always healthy! @baumanta and I had quite an in-depth discussion about this as a result actually.

I believe your suggestion does raise a point, but it is not correct per se. last_wp_pol is not meant to be a 'vector', much less one pointing from the last waypoint to the current position. Let me explain:
`last_sent_wp` is not a position we've actually ever reached. Maybe the term "waypoint" is confusing you here - it is merely the "carrot" in the "carrot and stick" analogy that we're playing with the PX4 position controller. The planner runs much faster than it would take the drone to actually travel that far.
![image](https://user-images.githubusercontent.com/14265408/54417425-73cf8f80-4702-11e9-9c8b-7d33d62d0350.png)

The way the smoothing works (and this is where I agree with you that it could (and should!) be done better) is that the direction to where the "carrot" was in the previous iteration. This is currently projected out to the goal distance, then this direction is taken into account when calculating the smoothing cost.

Instead, I could imagine that it would be better to take the current motion vector for this, or the polar representation of the previous carrot relative to the previous pose as an approximation thereof.

However, a vector from the previous carrot position to the current drone location is likely to point almost exactly backwards if where we are trying to go.

Reply:
Thanks for your explanation!

I got it, since the planner runs much faster, so the position of  "last_sent_wp" is still a "waypoint" the drone will be reached. Then "last_wp_pol" should be the direction of current position to "last_sent_wp".

And thanks again for your contribution.

Reply:
@sunzj I just opened [PR 252](https://github.com/PX4/avoidance/pull/252) for the changes @nicovanduijn mentioned. Our solution is correct, but maybe not the best one. The PR contains a visualization of the cost terms you might find interesting.